### PR TITLE
Bug/1108 non string fields

### PR DESCRIPTION
**Pull request by kzangeli**
Implements #1108 

Better error handling for non-string fields.

### CODE REVIEW

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test [hunk 1]
```diff
@@ -0,0 +1,194 @@
+# Copyright 2015 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: I understand that the restriction is in entity id and type, so probably we can do this renaming:

1108_non_string_fields/non_string_fields.test

to

1108_entity_id_and_type_non_string_fields/entity_id_and_type_non_string_fields.test

or something like that...

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+POST /v2/entities with invalid service-path
```
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: I seems that this NAME doesn't correspond with the purpose of the test...

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. POST /v2/entities with non-string in entity::id
+# 02. POST /v2/entities with non-string in entity::type
+# 03. POST /v2/entities with non-string in attribute::type
+# 04. POST /v2/entities with non-string in metadata::type
+# 05. POST /v2/entities with JSON object in metadata::value
```
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: This is going to change in the future when we allow JSON objects/vector for metadata value. Thus, I'd add a FIXME here about this step should be removed when time comes.

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 1 above)
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Looking closer to the .test, it seems that is not only a matter of entity::id and entity::type. Maybe my previous comment doesn't make sense, or the suggested name is not the right one.

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+POST /v2/entities with invalid service-path
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 01. POST /v2/entities with non-string in entity::id
```
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: I think that attribute::name and metadata::name should also should have similar restrictions. They are similar in nature to an entity::id, aren't they?

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+# You should have received a copy of the GNU Affero General Public License
+# along with Orion Context Broker. If not, see http://www.gnu.org/licenses/.
+#
+# For those usages not covered by this license please contact with
+# iot_support at tid dot es
+
+# VALGRIND_READY - to mark the test ready for valgrindTestSuite.sh
+
+--NAME--
+POST /v2/entities with invalid service-path
```
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: Wrong NAME?

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+--SHELL-INIT--
+dbInit CB
+brokerStart CB
+
+--SHELL--
+
+#
+# 05. POST /v2/entities with JSON object in metadata::value
```
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: Step should start at 01

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test (hunk 5 above)
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: Actually this test seems to be a sub-case of the previous one... Maybe I'm missing something.

File: CHANGES_NEXT_RELEASE [hunk 7]
```diff
@@ -15,3 +15,4 @@ Fix:  Checking for forbidden chars in many fields of JDON payload for API versio
 Fix:  Bug fix for receiving compound metadata values (which is not supported) (Issue #1110)
 Add:  GET /v2/entities/<entityId>/attrs/<attrName>/value - get a single attribute value (Issue #997)
 Add:  URI parameters id, idPattern and type for GET /v2/entities (Issues #952 #953 #969)
+Fix:  error handling for non-string entity::id/type (Issue #1108)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: Looking to the test, it seems to apply to additional fields.

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: In general, yes, of course.
However, the test is a part of another test (its name is non_string_fields_05 - meaning "part 05 of non_string_field).
I have used this for some time now ...

A test fails and I extract only the failing part, in a new functest.
Now that I have the new functest I can either remove it or keep it. I usually keep it

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: Yes, it is a sub-case (part 05) of non_string_fields_05.test

The name should change, of course. The name may no longer make sense.

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Cut 'n paste error ...

File: CHANGES_NEXT_RELEASE (hunk 7 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: True, it got a little bigger.
What's the best thing to do, add to the issue?  Create additional issues?

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 1 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Yeah, it's more like fields that should be strings but aren't

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Yes, I don't why I didn't add name for attribute/metadata ...
entity::id is just that ...   I don't remember.
Will add that check + functest covering it

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Sure, if you insist ... :-)
I am pretty sure we will notice the error in the test before we see the FIXME ;-)

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test (hunk 6 above)
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: I understand the value of "partical .test" during the debug steps, but once the functionality is ready at PR stage, I think is better to remove them, in order to keep the test base clean and maintanable.

File: CHANGES_NEXT_RELEASE (hunk 7 above)
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: The simplest way would be to add the other fields covered by the checking in this same line.

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 3 above)
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: You are right, but once the error happens in the future it is better to have a FIXME in the .test explaining why is failing to help us to realize about it (now it seems very simple to realize, but who knows in a couple of months...).

File: CHANGES_NEXT_RELEASE (hunk 7 above)
Comment (kzangeli) [inline CHANGES_NEXT_RELEASE]: Fixed in 6a46ebd

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 4 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Now I remember why there is no test for attribute::name and metadata::name.
In API v2, attributes and metadatas aren't like in v1:

    V2:
    "attr1" : { "type": "T1", "value": "" }

    V1:
    "attributes":
    {
        "name": "xxx", ...
    }

There is no need to check that attribute:name and metadata::name are strings. If they aren't, there will be a general JSON parser error (and also, no way for me to realize what they intended to do)

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 3 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Fixed in c787dda

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 2 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: Fixed in d7aafbf

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test (hunk 6 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: ok

Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: Fixed in 57de20f

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test (hunk 5 above)
Comment (kzangeli) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields_05.test]: Fixed in 57de20f

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 1 above)
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: NTC

File: test/functionalTest/cases/1108_non_string_fields/non_string_fields.test (hunk 4 above)
Comment (fgalan) [inline test/functionalTest/cases/1108_non_string_fields/non_string_fields.test]: NTC

(At lest in the scope of this PR... attribute:name and metadata:name may need checking in the normalized mode https://github.com/telefonicaid/fiware-orion/issues/1007, when we implemented that)

**Comment by fgalan** [comment]
LGTM

**Pull request merged**